### PR TITLE
Align static pipeline

### PR DESCRIPTION
**Pull request by as-suvorov**

### CODE REVIEW

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 1]
```diff
@@ -49,7 +23,7 @@ void fill_tensor(ov::Tensor tensor, T fill_val) {
 }
 
 template <typename T>
-void copy_to_tensor(const std::vector<T>& src_vec, ov::Tensor dst_tensor) {
+void copy_to_tensor(const std::vector<T>& src_vec, ov::Tensor&& dst_tensor) {
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: `&&` seems not needed there:
1) `dst_tensor` isn't supposed to be moved to this function
2) ov::Tensor itself is `pimpl` its copy to function should be free, so even `&` not needed

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 2]
```diff
@@ -73,7 +47,7 @@ ov::Tensor encode(ov::InferRequest& request,
 }
 
 // FIXME: Duplicate from llm_pipeline_static.cpp - need to reuse instead of copy-paste
-ov::Tensor make_tensor_slice(ov::Tensor tensor, size_t dim, size_t start_pos, size_t end_pos) {
+ov::Tensor make_tensor_slice(ov::Tensor& tensor, size_t dim, size_t start_pos, size_t end_pos) {
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: also, `&` not needed here

File: src/cpp/src/whisper_pipeline_static.hpp [hunk 3]
```diff
@@ -3,31 +3,26 @@
...
+#include "openvino/genai/whisper_pipeline.hpp"
 #include "whisper/whisper_models.hpp"
+#include "whisper_pipeline_base.hpp"
 
 namespace ov {
 namespace genai {
 
 class WhisperPipeline::StaticWhisperPipeline : public WhisperPipeline::WhisperPipelineImplBase {
 public:
-    StaticWhisperPipeline(const std::filesystem::path& model_path,
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.hpp:21]: 3 ctors removed, 2 added. Which one is redundant?

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 4]
```diff
@@ -130,43 +100,79 @@ void update_past_key_value(ov::InferRequest& source,
...
+    const size_t seq_length = input_ids_tensor.get_shape()[1];
+
+    OPENVINO_ASSERT(seq_length >= init_ids.size());
+
+    const size_t pad_size = seq_length - init_ids.size();
+
+    // pad right
+    // input_ids [token, token, token, pad_token]
+    // attention_mask [1, 1, 1, 0]
+    // normally pad left expected to work, but it didn't
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: I believe it works as expected, can be removed

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 5]
```diff
@@ -130,43 +100,79 @@ void update_past_key_value(ov::InferRequest& source,
...
+    auto input_ids_data = input_ids_tensor.data<int32_t>();
+    std::copy(init_ids.begin(), init_ids.end(), input_ids_data);
+    std::fill_n(input_ids_data + init_ids.size(), pad_size, static_cast<int32_t>(pad_token));
+
+    auto attention_mask_data = attention_mask_tensor.data<ov::float16>();
+    std::fill_n(attention_mask_data, init_ids.size(), 1u);
+    std::fill_n(attention_mask_data + init_ids.size(), pad_size, 0u);
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: hmm, shouldn't it be:
```
std::fill_n(attention_mask_data + init_ids.size(), init_ids.size() + pad_size, 0u);
```

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 6]
```diff
@@ -130,43 +100,79 @@ void update_past_key_value(ov::InferRequest& source,
...
+
+    const size_t pad_size = seq_length - init_ids.size();
+
+    // pad right
+    // input_ids [token, token, token, pad_token]
+    // attention_mask [1, 1, 1, 0]
+    // normally pad left expected to work, but it didn't
+    auto input_ids_data = input_ids_tensor.data<int32_t>();
+    std::copy(init_ids.begin(), init_ids.end(), input_ids_data);
+    std::fill_n(input_ids_data + init_ids.size(), pad_size, static_cast<int32_t>(pad_token));
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: And here as well??

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 7]
```diff
@@ -184,6 +190,8 @@ void zero_past_key_values(ov::InferRequest& request) {
 }
 
 void prepare_decoder_with_past(ov::InferRequest& decoder_with_past, ov::InferRequest& decoder) {
+    // initial mask size of 3 tested with 4 init tokens
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: if `init_ids` may have different size now, should we get rid of hardcoded `3` in this part?

File: src/cpp/src/whisper_pipeline_static.hpp (hunk 3 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.hpp:21]: 2 were removed, one added. Redundant was one with tokenizer.

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 1 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: done

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 2 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: done

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 4 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: removed

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 5 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: It seems correct to me. Anyway I rewritten this part to use `std::fill` as `pad_size` is not needed.

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 6 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: same as with `attention_mask`

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 7 above)
Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: It worked for me with hardcoded value during manual testing. I removed this comment

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 8]
```diff
@@ -197,10 +203,45 @@ void prepare_decoder_with_past(ov::InferRequest& decoder_with_past, ov::InferReq
...
+    int64_t output_token;
+    float max_prob = -std::numeric_limits<float>::infinity();
+
+    for (auto [_, lang_token] : config.lang_to_id) {
+        auto prob = logits_data[lang_token];
+        if (prob > max_prob) {
+            max_prob = prob;
+            output_token = lang_token;
+        }
+    }
```
Comment (as-suvorov) [inline src/cpp/src/whisper_pipeline_static.cpp:229]: FYI this part is different from stateful pipeline. There was an issue with `en` language detection so I suppressed non language tokens here. HF does same thing, so maybe I'll add it to the stateful pipeline as it more efficient.

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 9]
```diff
@@ -277,85 +331,143 @@ bool check_decoder_model_compatibility(const std::shared_ptr<ov::Model>& decoder
...
         }
 
-        DecodedResults decoded_results{std::vector{m_tokenizer.decode(output_tokens)}, std::vector{1.f}};
-        return decoded_results;
+        if (cancelled) {
+            break;
+        }
+    }
+
+    m_models.decoder_with_past.reset_state();
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp]: `decoder_with_past` not supposed to have state (at least for NPU-friendly models)

Comment (as-suvorov) [reply src/cpp/src/whisper_pipeline_static.cpp]: Removed

**Pull request merged**